### ISSUE TITLE
Batch Size 1 in training code

### ISSUE THREAD

**Issue by its-dron** [gtolias/how#6]
The paper refers to using a batch size of 5, and indeed a batch size of 5 is requested from the dataloader. 
However, the code explicitly iterates and backprops over each sample in the batch separately - effectively a batch size of 1.

The relevant code is here: https://github.com/gtolias/how/blob/master/how/stages/train.py#L91-L97

The ContrastiveLoss function used [supports multiple queries](https://github.com/filipradenovic/cnnimageretrieval-pytorch/blob/master/cirtorch/layers/loss.py#L10) - there's no reason to compute the loss per-tuple. Additionally, why does the model iterate over each image in the tuple instead of processing them as a batch?

**Comment by its-dron** [comment]
Apologies, I misread the indentation on where the optimizer step happened - https://github.com/gtolias/how/blob/master/how/stages/train.py#L99

I still think this is inefficient, but at least it's correct.

**Comment by jenicek** [comment]
Backpropagating for each tuple saves GPU memory as the computational graph is stored only for one tuple at a time. Images from a tuple are processed one by one as each image has a different size. This behavior is consistent with [training in cirtorch](https://github.com/filipradenovic/cnnimageretrieval-pytorch/blob/master/cirtorch/examples/train.py). If you have enough GPU memory or decrease the image size, you can straightforwardly adapt the code to minimize training time.
